### ISSUE TITLE
(R)Evolution of internal file formats

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#3087]
Working on RNA mass spectrometry, one problem is that there is no file format for storing oligonucleotide identification results. We have an experimental tool for doing database searching of MS2 spectra against RNA sequences and are working on a unified data structure that can represent peptide/protein, oligo/RNA or small molecule identification results (see #2935). It would be great to be able to reuse existing functionality e.g. from the IDFilter or FalseDiscoveryRate tools, but that would require a file format for passing the data around. I started working on an extended mzTab version with support for RNA, but there are problems with using that as an internal format:
- reading of mzTab isn't implemented
- there are some questionable design decisions in mzTab (see here: https://github.com/OpenMS/OpenMS/pull/2935#issuecomment-337939005)
- the "coarse" tabular format is restrictive and in many cases inefficient

Small molecule ID with OpenMS faces similar problems, although there is a proposal to extend idXML for that (#1659). However, given the various shortcomings of idXML, I'm convinced that the format needs to be rewritten or replaced.

But even when we get a new ID format that covers RNA and small molecules, these still cannot be properly represented together with quantitative data in featureXML/consensusXML. So we would need to make significant changes to these formats as well. There are additional problems with featureXML/consensusXML, too:
- no support for experimental design
- not all information from featureXML can be represented in consensusXML
- no support for "high-level" (e.g. peptide/protein) abundances

There are also general problems with XML-based file formats:
- they are often cumbersome to implement
- they are inefficient (disk space, I/O time/memory)
- they are hard to evolve (see e.g. #1671)

(Advantages of XML are that it's quasi-human readable, easily supports hierarchies and is very widely supported.)
To address some of these problems, @hroest and @grosenberger have started developing SQLite-based file formats for OpenSWATH.

I believe at this point we need to start thinking hard about the future of OpenMS' internal file formats. We have come to a point where we can't keep making small, incremental changes to them (most obviously in the case of idXML if we want to support RNA and small molecules), so we need to – but also have the opportunity to! – rethink things on a more basic level.
So, if we start from a blank slate, what technologies might be useful to create file formats for passing data between OpenMS/TOPP tools?
- XML (duh)
- SQLite
- Protocol Buffers
- Boost Serialization
- maybe others?

Based on limited research, I think Boost Serialization looks like a promising option, because:
- Boost is already a dependency
- it doesn't need a separate compilation step (unlike ProtoBuf)
- it supports different output (archive) formats, including XML
- it supports versioning
- importantly, it could be evolved together with the internal data structures

(The problem that I currently see with SQLite is that it would require a low-level implementation, i.e. hand-rolled ORM. There is however this interesting project that offers a Boost Serialization-like API for using SQLite from C++: https://github.com/paulftw/hiberlite)

To summarize, we need to make big changes to the file formats if we want proper support for small molecules and RNA (which I do want proper support for). I think we should use this opportunity to switch to a technology that will make it easier for us to deal with these file formats in the future.

**Comment by hroest** [comment]
This are some very good thoughts and I think it is important to start thinking about the future. We have done that in OpenSWATH quite a lot and we have come to some of the same conclusion, mainly that XML-based format are not very scalable, but they are good for archiving of data. Overall I agree with your assessment that it would be useful to have a file format that is more generic than the ones that we have and also can represent small molecule or RNA mass spec data. Also I assume you want to move forward and not wait for a standard committee to implement these changes into HUPO PSI formats (if they ever will). 
I also agree with your assessment of XML file formats, they do take a lot of space, are slow to read in general and are hard to read in a random-access fashion (I did implement a few things for doing this but this actually requires quite a lot of code and does not work well with compression). I think these points all compound each other, because they are verbose they need complicated parsers and are therefore slow to read. Also they dont deal well with binary data. I would like to add another contender to your list, namely hdf5 which is a great data format used in other disciplines and also used in some proteomics applications (e.g. mz5). 

One question is whether your proposal for a data structures is for temporary storage only or also for long-term archiving? I think there are very different requirements on either of these. 

Another question is that we dont just want to pass data around in OpenMS but also pass data around to other tools. I think this is very boost serialization becomes problematic as I dont really see how this could work across projects or programming languages. Also it seems that it couples the data format and the objects in memory very tightly, potentially preventing abstractions and locking us into certain data structures. I am afraid that if we go down that route then either (i) we have constant version changes of the format or (ii) we are locked into our data structures and cant change them any more without creating a whole host of problems. 

I see your point regarding SQLite as producing some extra work (e.g. we have to write some ORM code). The great thing about SQLite of course is that you can have part of your data on disk and you can query only part of it that you currently need. Again, this could be a very useful feature, especially for large data sets. On the other hand, of course querying is slow and should be done as little as possible. Still it may be feasible to iterate over thousands of files as once that would never fit into memory and one can select either all features from one file or select features across all files from a specific RT slice. Also if performance is ever an issue one could of course exchange SQLite with a more powerful system. I like hiberlite but I fear that we may run into the same issue as with boost serialization if we make the file format really a 1:1 representation of our data structures.

**Comment by timosachsenberg** [comment]
Thanks, @hendrikweisser for bringing this up for discussion and also thanks @hroest for the analysis. I am 100% in line with @hroest opinion and have been looking into different approaches from time to time.
I am currently also in favor of using SQLite and think that, at the current time, writing our own ORM code is the only solution. As long as we don't have proper reflections in C++ is not much difference anyway in terms of work that needs to be performed (but we get all the benefits from SQLite).

A bit of topic but related: What I am a bit more concerned with is if we manage to design our internal data structures in a way that we get a good OO class design and a nice ORM. For the new identification data structure @hendrikweisser proposed, it seems like we are moving from a class-based OO design to a more table-like design. Don't misunderstand this as critics, it just gives me a mixed feeling and I want to understand the implications better.

**Comment by hendrikweisser** [comment]
One thing I really like about Boost Serialization is that you should be able to get a binary file format (hopefully efficient) and an XML file format (interoperable) from the same implementation. Two birds with one stone. XML could still be used for archiving and working across frameworks.

> Also it seems that it couples the data format and the objects in memory very tightly, potentially preventing abstractions and locking us into certain data structures. I am afraid that if we go down that route then either (i) we have constant version changes of the format or (ii) we are locked into our data structures and cant change them any more without creating a whole host of problems.

Good point. This is a general question of having separate file format implementations (like now) vs. having "serialize"/"hibernate" functions in the data structures, somewhat independent of what file format technology (binary, XML, SQLite...) you use.
I think now we have the problem that we are locked into our file formats because we can't change them without breaking backwards compatibility. This also prevents us from making significant changes to the data structures. My hope would be that we can get a mechanism that allows us to evolve data structures and file formats together, with some support for backwards compatibility. I think the idea of having "fixed" file formats that you don't have to adapt is unrealistic. I'd rather we come to terms with the idea of "constant version changes" (which we have anyway) and find ways to handle them.

> I see your point regarding SQLite as producing some extra work (e.g. we have to write some ORM code). 
> I am currently also in favor of using SQLite and think that, at the current time, writing our own ORM code is the only solution.

I worry that this will be too much work and nobody will do it - especially because it's a different skillset from OOP. Looking e.g. at the sqMass implementation, the design decisions made in the ORM there are not trivial.
Adding "serialize" functions to data structures just seems much more feasible to me. It would also be easier to adapt when the data structures change.

> For the new identification data structure @hendrikweisser proposed, it seems like we are moving from a class-based OO design to a more table-like design.

Not so much with the intention of making it SQL-compatible, but a) to avoid redundancy and b) to allow lots of different entities to reference each other. @timosachsenberg, if you have suggestions to improve the "OO class design", please share them.

**Comment by timosachsenberg** [comment]
I can understand your point but I am very skeptical that this eases the evolution of our internal data formats. Boost serialization comes with support for different versions... but what does that mean? We also have different versions of e.g. idXML file format but never managed to implement the handling on top. You would need to write manual conversion functions between different version for every minor change (that's probably also a reason why we have a lot of key/value metadata ;) ).
In my view this only moves the effort from one part to the other without any net advantage.

See the arguments on e.g,: https://softwareengineering.stackexchange.com/questions/191269/java-serialization-advantages-and-disadvantages-use-or-avoid that promote serialization as a way to prototype persistence "For almost all other uses, the drawbacks you (and the article) mentions are too big: the exact format is hard to keep stable, class changes can easily make your serialized data unreadable, reading/writing the data in non-Java code is almost impossible (or at least a lot harder than necessary)"

ORM is quite some work to develop but I think residing to "direct" serialization is no real shortcut. If you want a reasonably fast and versatile way to store and retrieve your data, then there is still no way around databases I assume.

P.S.: I remember there has been an effort in OpenMS to serialize classes but it got discontinued many years ago. Unfortunately, I don't know the reasons why.

**Comment by hroest** [comment]
I agree that serialization can only be a solution either for prototyping or "caching" of data structures, e.g. to reduce some of the overhead that we have of passing our data structures from tool to tool (e.g. if we have a pipeline of 10 tools, then writing and reading XML files 10 times is some overhead). However, to be honest, to solve this problem, I have found that it is more efficient to just do it all in memory and write a single tool that keeps data in memory and somewhat abandon the concept of using 10 different tools. Also, when using workflow managers it adds complexity when there are 2 different ways to pass data between tools and we have to make sure to use the internal data structures for the internal part of the pipeline and then convert to "external" archival format. 

If used in this way, then we have not solved the fundamental problem that we still need a mapping of our internal data structure to some external archival format (be it idXML or mzIdentML). I dont see serialization as a solution for long-term storage (e.g. longer than a workflow runs) due to issues with compatibility with other software and future versions of OpenMS, it could actually be harmful if there is only a single version of OpenMS that can read a particular version of the data format. We always had the OpenMS internal file formats which were XML based and we basically had the worst of both worlds: a file format that is difficult / inefficient to parse but still not compatible with other tools. 

Another aspect to consider is that it may make sense to design a file format and put some work into it, whether its hdf5, SQLite or some other binary format, and simply "dumping" the object structure to disk to me seems really just a short-term solution for a very limited set of application (basically if you want to pass a structure in-memory but somehow cannot due to the structure of the problem) where other solutions would be better suited. To me serialization seems like an approach that solves the problem passing an object to some other part of the software through a disk / network interface instead of in-memory which is a very particular application that is not often required.

**Comment by timosachsenberg** [comment]
@hendrikweisser regarding the class design I currently can't come up with anything better what you have. Maybe going the road of explicitly modelling everything (maybe templated - to mitigate the type explosion) classes. Anyhow, if we want(ed) to provide an interface that e.g. lets you query all protein-protein cross-links, then this information will probably be returned in a nice structure that is created from the underlying maps. Same with adding a cross-link. We probably want to add them in a simple struct. So we get kind of class representation for the entities and the way you store it is basically a storage layer that is probably easily written to a database.

**Comment by hendrikweisser** [comment]
Thanks for the discussion, Hannes and Timo. @poshul and I have solved our urgent issue of getting FDRs for RNA search results by putting that functionality directly into the search engine, avoiding the file format problem (for now).
For the future, I agree to go with SQLite-based file formats, given that you both think it's the better approach. I have to finish some things for the `IdentificationData` class, but after that I would start working on developing an SQLite file format for it.
Alongside that, we (someone?) would need to develop an SQLite format for quantitative mass spec results. I would prefer to have one format that can store any/all of features, consensus features, and higher-level (e.g. peptide/protein) quantitative data. It should also be able to integrate identification results, by incorporating the `IdentificationData` schema from above.

**Comment by timosachsenberg** [comment]
Sounds great!

**Comment by hendrikweisser** [comment]
Closing this for being outdated.
